plugins/woocommerce/src/Admin/API/Init.php

```diff
@@ -54,115 +55,167 @@ public function __construct() {
+			'Automattic\WooCommerce\Admin\API\OnboardingProductTypes',
+			'Automattic\WooCommerce\Admin\API\OnboardingProfile',
+			'Automattic\WooCommerce\Admin\API\OnboardingTasks',
+			'Automattic\WooCommerce\Admin\API\OnboardingThemes',
+			'Automattic\WooCommerce\Admin\API\OnboardingPlugins',
+			'Automattic\WooCommerce\Admin\API\OnboardingProducts',
+			'Automattic\WooCommerce\Admin\API\MobileAppMagicLink',
+			'Automattic\WooCommerce\Admin\API\ShippingPartnerSuggestions',
+		);
 
-			if ( Features::is_enabled( 'analytics' ) ) {
-				$analytics_controllers = array(
-					'Automattic\WooCommerce\Admin\API\Customers',
-					'Automattic\WooCommerce\Admin\API\Leaderboards',
-					'Automattic\WooCommerce\Admin\API\Reports\Controller',
-					'Automattic\WooCommerce\Admin\API\Reports\Import\Controller',
-					'Automattic\WooCommerce\Admin\API\Reports\Export\Controller',
-					'Automattic\WooCommerce\Admin\API\Reports\Products\Controller',
-					'Automattic\WooCommerce\Admin\API\Reports\Variations\Controller',
-					'Automattic\WooCommerce\Admin\API\Reports\Products\Stats\Controller',
-					'Automattic\WooCommerce\Admin\API\Reports\Variations\Stats\Controller',
-					'Automattic\WooCommerce\Admin\API\Reports\Revenue\Stats\Controller',
-					'Automattic\WooCommerce\Admin\API\Reports\Orders\Controller',
-					'Automattic\WooCommerce\Admin\API\Reports\Orders\Stats\Controller',
-					'Automattic\WooCommerce\Admin\API\Reports\Categories\Controller',
-					'Automattic\WooCommerce\Admin\API\Reports\Taxes\Controller',
-					'Automattic\WooCommerce\Admin\API\Reports\Taxes\Stats\Controller',
-					'Automattic\WooCommerce\Admin\API\Reports\Coupons\Controller',
-					'Automattic\WooCommerce\Admin\API\Reports\Coupons\Stats\Controller',
-					'Automattic\WooCommerce\Admin\API\Reports\Stock\Controller',
-					'Automattic\WooCommerce\Admin\API\Reports\Stock\Stats\Controller',
-					'Automattic\WooCommerce\Admin\API\Reports\Downloads\Controller',
-					'Automattic\WooCommerce\Admin\API\Reports\Downloads\Stats\Controller',
-					'Automattic\WooCommerce\Admin\API\Reports\Customers\Controller',
-					'Automattic\WooCommerce\Admin\API\Reports\Customers\Stats\Controller',
-				);
-
-				// The performance indicators controllerq must be registered last, after other /stats endpoints have been registered.
-				$analytics_controllers[] = 'Automattic\WooCommerce\Admin\API\Reports\PerformanceIndicators\Controller';
-
-				$analytics_controllers = array_merge( $analytics_controllers, $analytic_mu_controllers );
-			}
-
-			$controllers = array_merge( $controllers, $analytics_controllers, $analytic_mu_controllers );
+		if ( ! did_filter( 'woocommerce_admin_rest_controllers' ) ) {
+			/**
+			 * Filter for the WooCommerce Admin REST controllers.
+			 *
+			 * @param array $controllers List of rest API controllers.
+			 *
+			 * @since 3.5.0
+			 */
+			$controllers = apply_filters( 'woocommerce_admin_rest_controllers', $controllers );
```

Comment:
Instead of conditionally calling a filter, but allowing all routes to be registered more than once, would it be make more sense to ensure `rest_api_init_wc_admin` is only called once? 

If its even needed. I wonder is this type of conditional filtering would cause race conditions with those consuming it.

Reply:
The issue is that `rest_api_init_wc_admin` previously ran once after the wc-admin and wc-analytics controllers were both added to the $controllers array. So it was added to both `::rest_api_init_wc_admin()` and `::rest_api_init_wc_analytics()`. Running this filter twice, though, could cause extensions to insert their own controllers twice.  

Looking through the existing extensions that use this filter, they all just add to the $controllers array.  So this was a way to make sure these would still be added as needed for either namespace, but only once.